src/music_gui.cpp

```diff
@@ -35,278 +36,388 @@
+{
+	if (this->selected_playlist == PLCH_CUSTOM1) return true;
+	if (this->selected_playlist == PLCH_CUSTOM2) return true;
+	return false;
+}
+
+/**
+ * Append a song to a custom playlist.
+ * Always adds to the currently active playlist.
+ * @param song_index Index of song in the current music set to add
+ */
+void MusicSystem::PlaylistAdd(size_t song_index)
 {
-	_music_wnd_cursong = 0;
-	DoStopMusic();
-	_song_is_active = false;
-	SetWindowWidgetDirty(WC_MUSIC_WINDOW, 0, 9);
+	if (!this->IsCustomPlaylist()) return;
+
+	/* Pick out song from the music set */
+	if (song_index >= this->music_set.size()) return;
+	PlaylistEntry entry = this->music_set[song_index];
+
+	/* Check for maximum length */
+	if (this->standard_playlists[this->selected_playlist].size() >= NUM_SONGS_PLAYLIST) return;
+
+	/* Add it to the appropriate playlist, and the display */
+	this->standard_playlists[this->selected_playlist].push_back(entry);
+	this->displayed_playlist.push_back(entry);
+
+	/* Add it to the active playlist, if playback is shuffled select a random position to add at */
+	if (this->active_playlist.empty()) {
+		this->active_playlist.push_back(entry);
+		if (this->IsPlaying()) this->Play();
+	} else if (this->IsShuffle()) {
+		/* Generate a random position between 0 and n (inclusive, new length) to insert at */
+		size_t maxpos = this->displayed_playlist.size();
+		size_t newpos = InteractiveRandom() % maxpos;
+		this->active_playlist.insert(this->active_playlist.begin() + newpos, entry);
+		/* Make sure to shift up the current playback position if the song was inserted before it */
+		if (newpos <= this->playlist_position) this->playlist_position++;
+	} else {
+		this->active_playlist.push_back(entry);
+	}
+
+	this->SaveCustomPlaylist(this->selected_playlist);
+
+	InvalidateWindowData(WC_MUSIC_TRACK_SELECTION, 0);
 }
 
-/** Begin playing the next song on the playlist */
-static void PlayPlaylistSong()
+/**
+ * Remove a song from a custom playlist.
+ * @param song_index Index in the custom playlist to remove.
+ */
+void MusicSystem::PlaylistRemove(size_t song_index)
 {
-	if (_cur_playlist[0] == 0) {
-		ResetPlaylist();
-		/* if there is not songs in the playlist, it may indicate
-		 * no file on the gm folder, or even no gm folder.
-		 * Stop the playback, then */
-		if (_cur_playlist[0] == 0) {
-			_song_is_active = false;
-			_music_wnd_cursong = 0;
-			_settings_client.music.playing = false;
-			return;
+	if (!this->IsCustomPlaylist()) return;
+
+	auto &pl = this->standard_playlists[this->selected_playlist];
+	if (song_index >= pl.size()) return;
+
+	/* Remove from "simple" playlists */
+	PlaylistEntry song = pl[song_index];
+	pl.erase(pl.begin() + song_index);
+	this->displayed_playlist.erase(this->displayed_playlist.begin() + song_index);
+
+	/* Find in actual active playlist (may be shuffled) and remove,
+	 * if it's the current song restart playback */
+	for (size_t i = 0; i < this->active_playlist.size(); i++) {
```

Comment:
could this loop be replaced with some incantation of `remove_if` ?

Reply:
I considered it, the problem is having to check the found index against playlist_position and restart playback in that case.